# <span style="color:rgb(160,0,86)">Zusammenhangsanalyse</span>

***

## <span style="color:rgb(160,0,86)">Lernziele</span>

- Sie verstehen den Sinn einer Zusammenhangsanalyse.
- Sie können eine Kontingenztafel lesen. 
- Sie können für nominale Daten den korrigierten Kontingenzkoeffizienten berechnen.
- Sie können den  und interpretieren.    

***

### <span style="color:rgb(160,0,86)">Was ist eine Zusammenhangsanalyse?</span>

Die Zusammenhangsanalyse ist ein Teilbereich der Statistik, das sich mit der Frage beschäftigt, ob und wie stark zwei oder mehr *Merkmale miteinander zusammenhängen*. Ziel ist es, **Beziehungen zwischen Variablen** zu erkennen, zu beschreiben und gegebenenfalls zu quantifizieren.

Je nach Art der Merkmale unterscheidet man verschiedene Verfahren:

- Für **nominale Merkmale** (z.B. Geschlecht, Rauchverhalten, usw.) können wir untersuchen, ob sich die relativen **Häufigkeiten unter den Merkmalen** unterscheiden.

- Für **ordinale Merkmale** (z.B. Schulnoten, Zufriedenheitsbewertungen, usw.) können wir zusätzlich prüfen, ob es einen *gerichteten Zusammenhang gibt*, die Merkmale miteinander steigen, entgegengesetzt verlaufen oder gar nichts miteinander zu tun haben. Ein **positiver Zusammenhang** besteht: Wenn ein Merkmal steigt, dann steigt auch das andere Merkmal. Ein **negativer Zusammenhang** besteht: Wenn ein Merkmal steigt, dann sinkt das andere Merkmal.

- Für **metrische Merkmale** (z.B. Körpergrösse, Gewicht, usw.) können wir zudem untersuchen, ob ein **linearer Zusammenhang**, also eine *Korrelation* besteht. Das ist möglich, wenn zahlenwertige Daten vorliegen, die auf  Zahlenachsen fest lokalisiert sind.  

<span style="color:rgb(160,0,86)">***Die Zusammenhangsanalyse ist wichtig, um Muster in Daten zu erkennen und fundierte Entscheidungen auf Basis von Daten zu treffen.***</span>

***

### <span style="color:rgb(160,0,86)">Wie untersuchen wir Zusammenhang in nominalen Daten?</span>

Eine **Kontingenztafel** ist eine tabellarische Darstellung, mit der man die **gemeinsame Häufigkeit** zweier Merkmale $\pmb{A}$ und $\pmb{B}$ zeigt. Sie hilft dabei, *Zusammenhänge* oder *Abhängigkeiten* zwischen diesen Merkmalen zu analysieren.

In dieser Tabel wird für jeden Wert $a_1, a_2, a_3, \ldots$ des Merkmals $\pmb{A}$ und für jeden Wert $b_1, b_2, b_3, \ldots$ des Merkmals $\pmb{B}$ angegeben, wie viele statistische Einheiten die Werte $a_i$ und $b_j$ haben: 

$$\begin{array}{c|cccc|c}
    & b_1 & b_2 & b_3 & \cdots & \\ \hline 
a_1 & n(a_1,b_1) & n(a_1,b_2) & n(a_1,b_3) & \cdots & n(a_1)\\
a_2 & n(a_2,b_1) & n(a_2,b_2) & n(a_2,b_3) & \cdots & n(a_2)\\
a_3 & n(a_3,b_1) & n(a_3,b_2) & n(a_3,b_3) & \cdots & n(a_3)\\
\vdots & \vdots & \vdots & \vdots & \ddots & \vdots\\ \hline
    & n(b_1) & n(b_2) & n(b_3) & \cdots & n
\end{array}$$

wobei $n(a_i)$ die **Zeilensummen**, $n(b_j)$ die **Spaltensummen** und $n$ die **Gesamtsumme** ist. Wenn wir alle diese absoluten Häufigkeiten durch $n$ dividieren, finden wir die **relativen Häufigkeiten**  

$$\begin{array}{c|cccc|c}
    & b_1 & b_2 & b_3 & \cdots & \\ \hline 
a_1 & h(a_1,b_1) & h(a_1,b_2) & h(a_1,b_3) & \cdots & h(a_1)\\
a_2 & h(a_2,b_1) & h(a_2,b_2) & h(a_2,b_3) & \cdots & h(a_2)\\
a_3 & h(a_3,b_1) & h(a_3,b_2) & h(a_3,b_3) & \cdots & h(a_3)\\
\vdots & \vdots & \vdots & \vdots & \ddots & \vdots\\ \hline
    & h(b_1) & h(b_2) & h(b_3) & \cdots & 1
\end{array}$$


Die zwei Merkmale $\pmb{A}$ und $\pmb{B}$ sind **unabhängig**, wenn
- die Verteilungen des Merkmals $\pmb{A}$ für alle Werte $b_1, b_2, b_3, \ldots$ mit der **Randverteilung** übereinstimmt: $$\frac{n(a_i,b_1)}{n(b_1)}=\frac{n(a_i,b_2)}{n(b_2)}=\frac{n(a_i,b_3)}{n(b_3)}=\ldots=\pmb{\frac{n(a_i)}{n}}=\pmb{h(a_i)}\;\,\text{ für alle Werte } a_i$$
- und die Verteilungen des Merkmals $\pmb{B}$ für alle Werte $a_1, a_2, a_3, \ldots$ mit der **Randverteilung** übereinstimmt: $$\frac{n(a_1,b_j)}{n(a_1)}=\frac{n(a_2,b_j)}{n(a_2)}=\frac{n(a_3,b_j)}{n(a_3)}=\ldots=\pmb{\frac{n(b_j)}{n}}=\pmb{h(b_j)}\;\,\text{ für alle Werte } b_j$$

### <span style="color:rgb(160,0,86)">Zum Beispiel</span>

Die folgenden zwei Merkmale $\pmb{A}$ und $\pmb{B}$ mit der Kontingenztafel
$$\begin{array}{c|ccccc|c}
    & b_1 & b_2 & b_3 & b_4 & b_5 &\\ \hline 
a_1 & 16 & 4 & 20 &  8 & 2 & 50\\
a_2 & 24 & 6 & 30 & 12 & 3 & 75\\ \hline
    & 40 & 10 & 50 & 20 & 5 & 125
\end{array}$$
sind **unabhängig**. Denn es gilt für die Verteilung von $\pmb{A}$
$$h(a_1)=\frac{16}{40}=\frac{4}{10}=\frac{20}{50}=\frac{8}{20}=\frac{2}{5}=\pmb{\frac{50}{125}}=\pmb{0.4}$$
$$h(a_2)=\frac{24}{40}=\frac{6}{10}=\frac{30}{50}=\frac{12}{20}=\frac{3}{5}=\pmb{\frac{75}{125}}=\pmb{0.6}$$
und für die Verteilung von $\pmb{B}$
$$h(b_1)=\frac{16}{50}=\frac{24}{75}=\pmb{\frac{40}{125}}=\pmb{0.32}$$
$$h(b_2)=\frac{4}{50}=\frac{6}{75}=\pmb{\frac{10}{125}}=\pmb{0.08}$$
$$h(b_3)=\frac{20}{50}=\frac{30}{75}=\pmb{\frac{50}{125}}=\pmb{0.40}$$
$$h(b_4)=\frac{8}{50}=\frac{12}{75}=\pmb{\frac{20}{125}}=\pmb{0.16}$$
$$h(b_5)=\frac{2}{50}=\frac{3}{75}=\pmb{\frac{5}{125}}=\pmb{0.04}$$
Wenn wir die **relativen Häufigkeiten** berechnen
$$\begin{array}{c|ccccc|c}
    & b_1 & b_2 & b_3 & b_4 & b_5 &\\ \hline 
a_1 & 0.128 & 0.032 & 0.160 &  0.064 & 0.016 & 0.4\\
a_2 & 0.192 & 0.048 & 0.240 & 0.096 & 0.024 & 0.6\\ \hline
    & 0.32 & 0.08 & 0.40 & 0.16 & 0.04 & 1
\end{array}$$
stellen wir fest, dass **bei unabhängigen Merkmalen** die relative Häufigkeit für die Werte $a_i$ und $b_j$ jeweils das **Produkt der Randverteilungen** ist:
$$\pmb{h(a_i,b_j)=h(a_i)\cdot h(b_j)} $$


In [1]:
from unittest.mock import inplace

import numpy as np

h_A = np.array([0.4, 0.6]).reshape(2, 1)
h_B = np.array([0.32, 0.08, 0.40, 0.16, 0.04]).reshape(1, 5)
h_A * h_B

array([[0.128, 0.032, 0.16 , 0.064, 0.016],
       [0.192, 0.048, 0.24 , 0.096, 0.024]])

Im Allgemeinen sind zwei Merkmale nicht unabhängig. Einen **extrem starken Zusammenhang** haben etwa die Merkmale $\pmb{C}$ und $\pmb{D}$ mit der Kontingenztafel 
$$\begin{array}{c|ccc|c}
    & d_1 & d_2 & d_3 &\\ \hline 
c_1 & 40 & 0 & 0 &  40\\
c_2 & 0 & 40 & 20 & 60\\ \hline
    & 40 & 40 & 20 & 100
\end{array}$$
Die Verteilung des Merkmals $\pmb{C}$ für die Werte $d_1$, $d_2$ und $d_3$ und die Verteilung des Merkmals $\pmb{D}$ für die Werte $c_1$ und $c_2$ sind nicht gleich wie die Randverteilungen. Wir können sogar aus dem Merkmal $\pmb{D}$ eindeutig auf das Merkmal $\pmb{C}$ schliessen.

***

Wie bei den **Masszahlen** zur Charakterisierung von der **Lage** und der **Streuung** *eines Merkmals*, können wir nun auch eine **Masszahl** für die **Unabhängigkeit** *zweier Merkmale* berechnen: 
- Zuerst bestimmen wir über alle Zeilen und Spalten der Kontingenztafel die Summe der quadrierten Pearson-Residuen $$\pmb{\chi^2 = \sum_{i,j}\frac{\big(h(a_i,b_j)-h(a_i)\cdot h(b_j)\big)^2}{h(a_i)\cdot h(b_j)}} $$
Wenn die Merkmale $\pmb{A}$ und $\pmb{B}$ **unabhängig** sind, dann ist $\pmb{\chi^2}=0$. Man kann zeigen, dass immer $$0\leqslant \chi^2 \leqslant \pmb{\min\{\text{Anzahl Zeilen},\text{Anzahl Spalten}\}-1} = \pmb{C_{\text{max}}}$$ gilt. Dieser Wert kann also für grosse Datensätze ($n$ sehr gross) selbst eine sehr grosse Zahl sein. Damit wir für jeden Datensatz maximal eine Eins erhalten, wird aus dieser **Masszahl für Unabhängigkeit** der folgende **korrigierte Kontigenzkoeffizient** definiert: $$\pmb{C_{\text{korr}}=\sqrt{1+\frac{1}{C_{\text{max}}}}\cdot\sqrt{\frac{\chi^2}{1+\chi^2}}} $$

- Wenn $\pmb{C_{\text{korr}}=0}$ gilt, dann sind die Merkmale $\pmb{A}$ und $\pmb{B}$ **unabhängig**.
- Wenn $\pmb{C_{\text{korr}}=1}$ gilt, dann sind die Merkmale $\pmb{A}$ und $\pmb{B}$ **maximal abhängig**. 

In [2]:
kontingenzTafel = np.array([[0.128, 0.032, 0.16, 0.064, 0.016],
                            [0.192, 0.048, 0.24, 0.096, 0.024]])

i = kontingenzTafel.shape[0]  # Anzahl Zeilen
j = kontingenzTafel.shape[1]  # Anzahl Spalten

C_max = min({i, j}) - 1
C_max

1

In [3]:
h_Arand = kontingenzTafel.sum(axis=1).reshape(i, 1)
h_Brand = kontingenzTafel.sum(axis=0).reshape(1, j)
rand = h_Arand * h_Brand
rand

array([[0.128, 0.032, 0.16 , 0.064, 0.016],
       [0.192, 0.048, 0.24 , 0.096, 0.024]])

In [4]:
chi_square = ((kontingenzTafel - rand) ** 2 / rand).sum()
chi_square

np.float64(4.814824860968089e-33)

In [5]:
C_korr = np.sqrt(1 + 1 / C_max) * np.sqrt(chi_square / (1 + chi_square))
np.round(C_korr, 4)

np.float64(0.0)

In [39]:
kontingenzTafel = np.array([[40, 0, 0],
                            [0, 40, 20]])
kontingenzTafel = kontingenzTafel / kontingenzTafel.sum()
kontingenzTafel

array([[0.4, 0. , 0. ],
       [0. , 0.4, 0.2]])

In [6]:
i = kontingenzTafel.shape[0]  # Anzahl Zeilen
j = kontingenzTafel.shape[1]  # Anzahl Spalten

C_max = min({i, j}) - 1
C_max

1

In [41]:
h_Arand = kontingenzTafel.sum(axis=1).reshape(i, 1)
h_Brand = kontingenzTafel.sum(axis=0).reshape(1, j)
rand = h_Arand * h_Brand
rand

array([[0.16, 0.16, 0.08],
       [0.24, 0.24, 0.12]])

In [8]:
chi_square = ((kontingenzTafel - rand) ** 2 / (rand)).sum()
chi_square

np.float64(4.814824860968089e-33)

In [9]:
C_korr = np.sqrt(1 + 1 / C_max) * np.sqrt(chi_square / (1 + chi_square))
np.round(C_korr, 4)

np.float64(0.0)

### <span style="color:rgb(160,0,86)">Zum Beispiel</span>

Es wurden 100 Personen zu den Merkmale **Geschlecht** und **Sporttreiben** befragt:  
$$\begin{array}{c|cc}
    & \text{ja} & \text{nein} &\\ \hline 
\text{Mann} & 30 & 20 \\
\text{Frau} & 25 & 25 \\ 
\end{array}$$

Sind die beiden Merkmale unabhängig?

In [10]:
kontingenzTafel = np.array([[33, 17],
                            [21, 29]])
kontingenzTafel = kontingenzTafel / kontingenzTafel.sum()
kontingenzTafel

array([[0.33, 0.17],
       [0.21, 0.29]])

In [11]:
i = kontingenzTafel.shape[0]  # Anzahl Zeilen
j = kontingenzTafel.shape[1]  # Anzahl Spalten

C_max = min({i, j}) - 1
C_max

1

In [12]:
h_Arand = kontingenzTafel.sum(axis=1).reshape(i, 1)
h_Brand = kontingenzTafel.sum(axis=0).reshape(1, j)
rand = h_Arand * h_Brand
rand

array([[0.27, 0.23],
       [0.27, 0.23]])

In [13]:
chi_square = ((kontingenzTafel - rand) ** 2 / (rand)).sum()
chi_square

np.float64(0.05797101449275362)

In [14]:
C_korr = np.sqrt(1 + 1 / C_max) * np.sqrt(chi_square / (1 + chi_square))
np.round(C_korr, 4)

np.float64(0.331)

Der **korrigierte Kontigenzkoeffizient** $\pmb{C_{\text{korr}}=0.331}$ liegt unter der Mittes des Intervalls $[0,1]$ und besagt somit, dass zwischen den beiden Markmalen **Geschlecht** und **Sporttreiben** im Datensatz ein **schwacher Zusammenhang** besteht.

### <span style="color:rgb(160,0,86)">Aufgabe 1</span>

In einer Umfrage ergibt sich für die Merkmale **Bruttogehalt** und **Geschlecht** die folgende Kontingenztafel:

$$\begin{array}{l|cc}
    & \text{Mann} & \text{Frau} &\\ \hline 
\text{0- 2000} & 0.1 & 0.05 \\
\text{2000- 4000} & 0.1 & 0.1 \\
\text{4000- 6000} & 0.2 & 0.2 \\
\text{6000-10000} & 0.2 & 0.05
\end{array}$$

- Sind die beiden Merkmale **Bruttogehalt** und **Geschlecht** unabhängig?
- Prüfen Sie mit einem **Hypothesentest** auf dem Signifikanzniveau $\alpha=5\%$, ob die Unterschiede zwischen Mann und Frau in dieser Zufallsstichprobe signifikant sind.   

In [129]:
kontingenzTafel = np.array([[0.1, 0.05], [0.1, 0.1], [0.2, 0.2], [0.2, 0.05]])


def calc_c_korr(c_table):
    shape_max = min(c_table.shape) - 1

    rand_a = c_table.sum(axis=1).reshape(c_table.shape[0], 1)
    rand_b = c_table.sum(axis=0).reshape(1, c_table.shape[1])

    rand_full = rand_a * rand_b

    chi2 = ((c_table - rand_full) ** 2 / rand_full).sum()
    c_korr = np.round(np.sqrt(1 + 1 / shape_max) * np.sqrt(chi2 / (1 + chi2)), 4)

    print("C Max: " + str(shape_max))
    print("Chi Square: " + str(chi2))
    print("C Korr: " + str(c_korr))


calc_c_korr(kontingenzTafel)

C Max: 1
Chi Square: 0.06944444444444443
C Korr: 0.3604


### <span style="color:rgb(160,0,86)">Aufgabe 2</span>

Im Datensatz *2021_Personalerhebung.csv* soll analysiert werden, ob die beiden Merkmale **Qualifikation** und **Bereichszugehörigkeit** unabhängig sind. 
- Bestimmen Sie den **korrigierten Kontingenzkoeffizienten** für die zwei Merkmale.

Die Werte der zwei Merkmale sollen klassiert werden. Im Merkmal **Qualifikation** wird nur noch zwischen *Akademiker* ("Bachelor", "Master", "Promotion") und *nicht Akademiker* ("Mittlere Reife", "Abitur") unterschieden. Im Merkmal **Bereichszugehörigkeit** werden die Abteilungen "Geschäftsführung" und "Finanzen" zum Wert *Bereich I*, die Abteilungen "Entwicklung" und "Test/Anwendungen" zum Wert *Bereich II* und die Abteilungen "Schulung" und "Vertrieb" zum Wert *Bereich II* zusammengefasst.
- Bestimmen Sie den **korrigierten Kontingenzkoeffizienten** für die zwei klassierten Merkmale.

In [132]:
import pandas as pd
import scipy.stats as stats
import numpy as np

df = pd.read_csv("Daten/2021_Personalerhebung.csv", delimiter=";")
df["Abteilung"] = df["Abteilung"].replace({
    "Geschaftsführung": "I",
    "Finanzen": "I",
    "Entwicklung": "II",
    "Test/Anwendungen": "II",
    "Schulung": "III",
    "Vertrieb": "III"
})

df["Ausbildung"] = df["Ausbildung"].replace({
    "Bachelor": "Akademiker",
    "Master": "Akademiker",
    "Promotion": "Akademiker",
    "Mittlere Reife": "Nicht Akademiker",
    "Abitur": "Nicht Akademiker"
})

contingency_table = pd.crosstab(df["Abteilung"], df["Ausbildung"]).to_numpy()
contingency_table = contingency_table / contingency_table.sum()
calc_c_korr(contingency_table)


C Max: 1
Chi Square: 0.12067099567099568
C Korr: 0.4641


### <span style="color:rgb(160,0,86)">Aufgabe 3</span>

Ein Informatik-Team entwickelt eine E-Learning-Plattform, auf der Studierende verschiedene Kurse belegen können. Um das Nutzerverhalten besser zu verstehen, wird untersucht, ob ein Zusammenhang zwischen der **Nutzungsdauer pro Woche** und der gewählten **Kursart** besteht.

Dazu wurden Daten von 400 Nutzern erhoben:

$$\begin{array}{l|ccccc}
& \text{\small Programmierung}	& \text{\small Mathematik}	& \text{\small Datenanalyse}	& \text{\small Soft Skills}	& \text{\small Projektmanagement}\\ \hline 
\text{\small < 2 h} & 10 & 20 & 25 & 30 & 15 \\
\text{\small 2–5 h} & 30 & 25 & 20 & 15 & 10 \\
\text{\small 6–10 h} & 35 & 20 & 15 & 10 & 20 \\
\text{\small > 10 h} & 25 & 15 & 20 & 15 & 25
\end{array}$$
- Bestimmen Sie den **korrigierten Kontingenzkoeffizienten** für die zwei Merkmale. Gibt es einen Zusammenhang zwischen der **Nutzungsdauer pro Woche** und der gewählten **Kursart**?

In [133]:
contingency_table = np.array([[10, 20, 25, 30, 15], [30, 25, 20, 15, 10], [35, 20, 15, 10, 20], [25, 15, 20, 15, 25]])
contingency_table = contingency_table / contingency_table.sum()
calc_c_korr(contingency_table)

C Max: 3
Chi Square: 0.09749999999999999
C Korr: 0.3442


### <span style="color:rgb(160,0,86)">Aufgabe 4</span>

Ein Softwareunternehmen möchte untersuchen, ob es einen Zusammenhang zwischen der **Nutzergruppe** und dem **bevorzugten Betriebssystem** gibt. Dazu wurden 500 Nutzer einer Systemumfrage ausgewertet.

$$\begin{array}{l|ccccc}
& \text{\small Windows}	& \text{\small macOS}	& \text{\small Linux}	& \text{\small ChromeOS}	& \text{\small Sonstige}\\ \hline 
\text{\small Privat} & 90 & 20 & 10 & 20 & 10 \\
\text{\small Beruflich} & 60 & 30 & 10 & 5 & 5 \\
\text{\small Studierende} & 40 & 30 & 20 & 30 & 10 \\
\text{\small Entwickler} & 10 & 10 & 50 & 10 & 30
\end{array}$$
- Bestimmen Sie den **korrigierten Kontingenzkoeffizienten** für die zwei Merkmale. Gibt es einen Zusammenhang zwischen der **Nutzergruppe** und dem **bevorzugten Betriebssystem**?

In [134]:
contingency_table = np.array([[90, 20, 10, 20, 10], [60, 30, 10, 5, 5], [40, 30, 20, 30, 10], [10, 10, 50, 10, 30]])
contingency_table = contingency_table / contingency_table.sum()
calc_c_korr(contingency_table)

C Max: 3
Chi Square: 0.3555216138432922
C Korr: 0.5914


![HSLU](Bilder/LogoHSLU.png)